crates/astria-sequencer/src/mempool/mod.rs

```diff
@@ -267,34 +286,104 @@ impl Mempool {
     }
 
     /// Updates stored transactions to reflect current blockchain state. Will remove transactions
-    /// that have stale nonces and will remove transaction that are expired.
+    /// that have stale nonces or are expired. Will also shift transation between pending and
+    /// parked to relfect changes in account balances.
     ///
     /// All removed transactions are added to the CometBFT removal cache to aid with CometBFT
     /// mempool maintenance.
     #[instrument(skip_all)]
-    pub(crate) async fn run_maintenance<F, O>(&self, current_account_nonce_getter: F)
-    where
-        F: Fn([u8; 20]) -> O,
-        O: Future<Output = anyhow::Result<u32>>,
+    pub(crate) async fn run_maintenance<F1, O1, F2, O2>(
+        &self,
+        current_account_nonce_getter: F1,
+        current_account_balances_getter: F2,
+    ) where
+        F1: Fn([u8; 20]) -> O1,
+        O1: Future<Output = anyhow::Result<u32>>,
+        F2: Fn([u8; 20]) -> O2,
+        O2: Future<Output = anyhow::Result<HashMap<IbcPrefixed, u128>>>,
     {
         let (mut pending, mut parked) = self.acquire_both_locks().await;
-
-        // clean accounts of stale and expired transactions
-        let mut removed_txs = pending.clean_accounts(&current_account_nonce_getter).await;
-        removed_txs.append(&mut parked.clean_accounts(&current_account_nonce_getter).await);
-
-        // run promotion logic in case transactions not in this mempool advanced account state
-        let to_promote = parked.find_promotables(&current_account_nonce_getter).await;
-        // Release the lock asap.
-        drop(parked);
-        for (ttx, current_account_nonce) in to_promote {
-            if let Err(error) = pending.add(ttx, current_account_nonce) {
-                error!(
-                    current_account_nonce,
-                    "failed to promote transaction during maintenance: {error:#}"
-                );
+        let mut removed_txs = Vec::<([u8; 32], RemovalReason)>::new();
+
+        // To clean we need to:
+        // 1.) remove stale and expired transactions from both pending and parked
+        // 2.) check if we have transactions in pending which need to be demoted due
+        //     to balance decreases
+        // 3.) if there were no demotions, check if parked has transactions we can
+        //     promote
+
+        // TODO reprice transactions also if FeeAssetChange or FeeChange Actions were ran :0
```

Comment:
Is this ticketed?

Reply:
it's written in: https://github.com/astriaorg/astria/pull/1411